### PR TITLE
feat(exporter/sumologic): add metrics for HTTP requests

### PR DESCRIPTION
**Pull request by sumo-drosiek**
Signed-off-by: Dominik Rosiek <drosiek@sumologic.com>

Generated metrics
```
# HELP otelcol_sumologic_requests_bytes Total size of requests (in bytes)
# TYPE otelcol_sumologic_requests_bytes counter
otelcol_sumologic_requests_bytes{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="0"} 297107
otelcol_sumologic_requests_bytes{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="200"} 1.727096e+06
# HELP otelcol_sumologic_requests_duration Duration of HTTP requests (in milliseconds)
# TYPE otelcol_sumologic_requests_duration counter
otelcol_sumologic_requests_duration{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="0"} 0
otelcol_sumologic_requests_duration{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="200"} 280
# HELP otelcol_sumologic_requests_records Total size of requests (in number of records)
# TYPE otelcol_sumologic_requests_records counter
otelcol_sumologic_requests_records{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="0"} 16384
otelcol_sumologic_requests_records{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="200"} 95232
# HELP otelcol_sumologic_requests_sent Number of requests
# TYPE otelcol_sumologic_requests_sent counter
otelcol_sumologic_requests_sent{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="0"} 2
otelcol_sumologic_requests_sent{endpoint="http://localhost:3000",pipeline="metrics",service_instance_id="00e0fe04-5899-4b4b-9e4d-c2dee53af447",service_version="latest",status_code="200"} 12
```

### CODE REVIEW

File: pkg/exporter/sumologicexporter/internal/observability/observability.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+func RecordRequestsSent(status_code int) {
+	stats.RecordWithTags(
+		context.Background(),
+		[]tag.Mutator{tag.Insert(statusKey, fmt.Sprintf("%d", status_code))},
+		mRequestsSent.M(int64(1)),
+	)
+}
+
+// RecordPodAdded increments the metric that records pod add events receiver.
+func RecordRequestsDuration(duration time.Duration, status_code int) {
```
Comment (sumo-drosiek) [inline pkg/exporter/sumologicexporter/internal/observability/observability.go:72]:
```suggestion
func RecordRequestsDuration(duration time.Duration, statusCode int) {
```

File: pkg/exporter/sumologicexporter/internal/observability/observability.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+// TODO: re-think if processor should register it's own telemetry views or if some other
+// mechanism should be used by the collector to discover views from all components
+
```
Comment (sumo-drosiek) [inline pkg/exporter/sumologicexporter/internal/observability/observability.go:29]:
```suggestion
```

File: pkg/exporter/sumologicexporter/sender.go [hunk 3]
```diff
@@ -57,6 +58,68 @@ type logPair struct {
...
+
+func (c *countingReader) withBytes(data []byte) *countingReader {
+	c.reader = bytes.NewReader(data)
+	return c
+}
+
+func (c *countingReader) withString(data string) *countingReader {
+	c.reader = strings.NewReader(data)
+	return c
+}
```
Comment (pmalek-sumo) [inline pkg/exporter/sumologicexporter/sender.go:82]: * I have a feeling that this whole section is a bit underdocumented.
* Additionally `countingReader` doesn't really count anything, does it? It just holds stuff

Comment (sumo-drosiek) [reply pkg/exporter/sumologicexporter/sender.go:82]: > Additionally countingReader doesn't really count anything, does it? It just holds stuff

would `readerState` be better choice

File: pkg/exporter/sumologicexporter/sender.go [hunk 4]
```diff
@@ -71,11 +71,13 @@ func newCountingReader(records int) *countingReader {
 	}
 }
 
+// withBytes set ups reader to read from bytes data
```
Comment (pmalek-sumo) [inline pkg/exporter/sumologicexporter/sender.go:74]:
```suggestion
// withBytes sets up reader to read from bytes data
```

File: pkg/exporter/sumologicexporter/sender.go [hunk 5]
```diff
@@ -57,6 +58,68 @@ type logPair struct {
...
+	return err
+}
+
+func (b *bodyBuilder) Len() int {
+	return b.builder.Len()
+}
+
+// toCountingReader converts bodyBuilder to countingReader
+func (b *bodyBuilder) toCountingReader() *countingReader {
+	return newCountingReader(b.counter).withString(b.builder.String())
```
[the remark this answers is not in the document]
Comment (pmalek-sumo) [reply pkg/exporter/sumologicexporter/sender.go:126]: Nvm, I thought there's a `bytes.Builder`

**Review by swiatekm-sumo — COMMENTED** [review]
If we're going to follow the OT core convention, then the names should be something like `otelcol_exporter_requests_count` with `exporter=<exporter_name>` as a tag.

I'm also wondering if we need any of these to be histograms. I'm not sure that the average size of a request - which is all we can get from totals - is going to be enough. I'd definitely like a histogram for durations.

**Comment by sumo-drosiek** [comment]
Updated list of metrics
```
# HELP otelcol_exporter_requests_bytes Total size of requests (in bytes)
# TYPE otelcol_exporter_requests_bytes counter
otelcol_exporter_requests_bytes{endpoint="http://localhost:3000",exporter="sumologic/custom",pipeline="metrics",service_instance_id="09c71af2-a80d-409c-af1f-3a20e9c039b5",service_version="latest",status_code="200"} 297086
# HELP otelcol_exporter_requests_duration Duration of HTTP requests (in milliseconds)
# TYPE otelcol_exporter_requests_duration counter
otelcol_exporter_requests_duration{endpoint="http://localhost:3000",exporter="sumologic/custom",pipeline="metrics",service_instance_id="09c71af2-a80d-409c-af1f-3a20e9c039b5",service_version="latest",status_code="200"} 43
# HELP otelcol_exporter_requests_records Total size of requests (in number of records)
# TYPE otelcol_exporter_requests_records counter
otelcol_exporter_requests_records{endpoint="http://localhost:3000",exporter="sumologic/custom",pipeline="metrics",service_instance_id="09c71af2-a80d-409c-af1f-3a20e9c039b5",service_version="latest",status_code="200"} 16384
# HELP otelcol_exporter_requests_sent Number of requests
# TYPE otelcol_exporter_requests_sent counter
otelcol_exporter_requests_sent{endpoint="http://localhost:3000",exporter="sumologic/custom",pipeline="metrics",service_instance_id="09c71af2-a80d-409c-af1f-3a20e9c039b5",service_version="latest",status_code="200"} 2
```

**Review by swiatekm-sumo — APPROVED** [review]

**Pull request merged**